language/move-lang/src/cfgir/liveness/mod.rs

```diff
@@ -0,0 +1,260 @@
+    let mut liveness = Liveness::new(reverse);
+    let final_invariants = liveness.analyze_function(reverse, initial_state).unwrap();
+    (final_invariants, liveness.states)
+}
+
+fn build_initial_states(cfg: &BlockCFG, final_invariants: FinalInvariants) -> InitialStates {
+    let mut states = BTreeMap::new();
+    for lbl in cfg.blocks().keys() {
+        let mut state = LivenessState::initial();
+        for predecessor in cfg.predecessors(*lbl) {
+            // get the final liveness results for each predecessor, and union them
+            if let Some(pre) = final_invariants.get(predecessor) {
+                state.extend(pre)
+            }
+        }
+        states.insert(*lbl, state);
+    }
+    states
+}
+
+fn command(state: &mut LivenessState, sp!(_, cmd_): &Command) {
+    use Command_ as C;
+    match cmd_ {
+        C::Assign(ls, e) => {
+            lvalues(state, ls);
+            exp(state, e);
+        }
+        C::Mutate(el, er) => {
+            exp(state, er);
+            exp(state, el)
+        }
+        C::Return(e) | C::Abort(e) | C::IgnoreAndPop { exp: e, .. } | C::JumpIf { cond: e, .. } => {
+            exp(state, e)
+        }
+
+        C::Jump(_) => (),
+    }
+}
+
+fn lvalues(state: &mut LivenessState, ls: &[LValue]) {
+    ls.iter().for_each(|l| lvalue(state, l))
+}
+
+fn lvalue(state: &mut LivenessState, sp!(_, l_): &LValue) {
+    use LValue_ as L;
+    match l_ {
+        L::Ignore => (),
+        L::Var(v, _) => {
+            state.live.remove(v);
+            state.assigned.insert(v.clone());
+        }
+        L::Unpack(_, _, fields) => fields.iter().for_each(|(_, l)| lvalue(state, l)),
+    }
+}
+
+fn exp(state: &mut LivenessState, parent_e: &Exp) {
+    use UnannotatedExp_ as E;
+    match &parent_e.exp.value {
+        E::Unit | E::Value(_) | E::UnresolvedError => (),
+
+        E::BorrowLocal(_, var) | E::Copy { var, .. } => use_local(state, var),
+
+        E::Move { var, .. } => {
+            use_local(state, var);
+            state.moved.insert(var.clone());
+        }
+
+        E::ModuleCall(mcall) => exp(state, &mcall.arguments),
+        E::Builtin(_, e)
+        | E::Freeze(e)
+        | E::Dereference(e)
+        | E::UnaryExp(_, e)
+        | E::Borrow(_, e, _) => exp(state, e),
+
+        E::BinopExp(e1, _, e2) => {
+            exp(state, e1);
+            exp(state, e2)
+        }
+
+        E::Pack(_, _, fields) => fields.iter().for_each(|(_, _, e)| exp(state, e)),
+
+        E::ExpList(es) => es.iter().for_each(|item| exp_list_item(state, item)),
+    }
+}
+
+fn exp_list_item(state: &mut LivenessState, item: &ExpListItem) {
+    match item {
+        ExpListItem::Single(e, _) | ExpListItem::Splat(_, e, _) => exp(state, e),
+    }
+}
+
+fn use_local(state: &mut LivenessState, local: &Var) {
+    state.live.insert(local.clone());
+}
+
+//**************************************************************************************************
+// Refinement
```

Comment:
Is "refinement" code transformation?

Reply:
I have two major passes in cfgir/mod.rs `verify` and `refine`. 

Originally `refine` was called `optimize`, but people voted that name down as it wasn't always really an "optimization"

Reply:
This is more of name bikeshedding for another time, but I would find `rewrite` or `transform` clearer. "Refine" has a specific meaning in logic/analysis, which makes it somewhat of a confusing choice for this phase.

Reply:
I would find `transform` and `rewrite` to be too vague. The thing being refined here is the code, not a context. I'll keep it as is for this PR, but am happy to discuss it later :)

Reply:
In short, I need a word close to but different than "optimize"